core/src/main/scala/org/apache/spark/scheduler/EventLoggingListener.scala

```diff
@@ -169,6 +182,31 @@ private[spark] class EventLoggingListener(
 
   // Events that trigger a flush
   override def onStageCompleted(event: SparkListenerStageCompleted): Unit = {
+    if (shouldLogExecutorMetricsUpdates) {
+      // clear out any previous attempts, that did not have a stage completed event
+      val prevAttemptId = event.stageInfo.attemptNumber() - 1
+      for (attemptId <- 0 to prevAttemptId) {
+        liveStageExecutorMetrics.remove((event.stageInfo.stageId, attemptId))
+      }
+
+      // log the peak executor metrics for the stage, for each live executor,
+      // whether or not the executor is running tasks for the stage
+      val accumUpdates = new ArrayBuffer[(Long, Int, Int, Seq[AccumulableInfo])]()
+      val executorMap = liveStageExecutorMetrics.remove(
+        (event.stageInfo.stageId, event.stageInfo.attemptNumber()))
+      executorMap.foreach {
+       executorEntry => {
+          for ((executorId, peakExecutorMetrics) <- executorEntry) {
+            val executorMetrics = new ExecutorMetrics(-1, peakExecutorMetrics.metrics)
```

Comment:
We need to pass in a value for timestamp, but there isn't really one for the peak metrics, since times for each peak could be different. 

When processing, -1 will help indicate that the event is coming from the history log, and contains the peak values for the stage that is just ending. When updating the stage executor peaks (peak executor values stored for each active stage), we can replace all of the peak executor metric values instead of updating with the max of current and new values for each metric. 

As an example, suppose there is the following scenario:
T1: start of stage 1
T2: peak value of 1000 for metric m1
T3: start of stage 2
T4: stage 1 ends, and peak metric values for stage 1 are logged, including m1=1000
T5: stage 2 ends, and peak metric values for stage 2 are logged.

If values for m1 are < 1000 between T3 (start of stage 2) and T5 (end of stage 2), and say that the highest value for m1 during that period is 500, then we want the peak value for m1 for stage 2 to show as 500.

There would be an ExecutorMetricsUpdate event logged (and then read) at T4 (end of stage 1), with m1=1000, which is after T3 (start of stage 2). If when reading the history log, we set the stage 2 peakExecutorMetrics to the max of current or new values from ExecutorMetricsUpdate, then the value for stage 2 would remain at 1000. However, we want it to be replaced by the value of 500 when it gets the ExecutorMetricsUpdate logged at T5 (end of stage 2). During processing of ExecutorMetricsUpdate, for the stage level metrics, it will replace all the peakExecutorMetrics if timestamp is -1.

I can add some comments for this.

Reply:
hmm, now I'm confused, sorry I am probably rehashing something we have discussed before.  What do you plan on replacing that -1 with later on?

My understanding was that you'd always keep one timestamp for the last time you detected a new max, for any metric.  You'd update the metrics per-stage, iff you were still within that stage.  so you'd lose the ability to tell in more detail what the max was for one stage in isolation, vs. stages running together (eg. if there was a max at T3.5, you'd just log that one, and so you'd lose info about the T2 max which was for stage 1 by itself).

Reply:
This is confusing, especially since the current code does not have the stage level metrics, just executor level. 

The -1 wouldn't be replaced. PeakExecutorMetrics only tracks the peak metric values for each executor (and later each executor per stage) and doesn't have a timestamp.

If there is a local max (500), which is the max between T3 and T5, it would be logged at time T5, even if it happens at T3.5.

In actual event order, what the driver sees when the application is running:
T1: start of stage 1
T2: value of 1000 for metric m1
T3: start of stage 2
T3.5: peak value of 500 for metric m1
T4: stage 1 ends
T5: stage 2 ends

Suppose that 1000 (seen at T2) is the peak value of m1 between T1 and T4, so it is the peak value seen while stage 1 is running. The m1=1000 value will be dumped as the max value in an executorMetricsUpdate event right before the stage 1 end event is logged. Also suppose that 500 (seen at T3.5 is the peak value of m1 between T3 and T5, so it is the peak value seen while stage 2 is running. The m1=500 value will be dumped as the max value in an executorMetricsUpdate right before the stage 2 end event is logged.

The generated Spark history log would have the following order of events:

Start of stage 1
Start of stage 2
executorMetricsUpdate with m1=1000
end of stage 1
executorMetricsUpdate with m1=500
end of stage 2

When the Spark history server is reading the log, it will will create the peakExecutorMetrics for stage 2 when stage 2 starts, which is before it sees the executorMetricsUpdate with m1=1000, and so will store m1=1000 as the current peak value. When it later sees the executorMetricsUpdate with m1=500, it needs to overwrite the m1 value (and set to 500), not compare and update to the max value (which would be 1000). The -1 would indicate that the event is coming from the Spark history log, is a peak value for the stage just about to complete, and should overwrite any previous values.

If the timestamp is set to a positive value, then it will do a compare and update to the max value instead.

Reply:
I can see how this would work, but it also seems far more confusing than necessary.  My understanding was that you'd always log the last timestamp which replaced the peak value for *any* metric.  Are you ever logging something other than -1 for the timestamp?  If not, we just shouldn't put any timestamp in the log.

It might be helpful to step back a bit , and rather than focusing on the mechanics of what you're doing now, discuss the desired end behavior in the history server and the live UI based on the timestamp.

Reply:
The last timestamp seems like it wouldn't have enough information, since peaks for different metrics could occur at different times, and with different combinations of stages running. 

Only -1 would be logged. Right now it's writing out SparkListenerExecutorMetricsUpdate events, which contain ExecutorMetrics, which has timestamp. Do you think timestamp should be removed from ExecutorMetrics? It seems good to have the timestamp for when the metrics were gathered, but it's not being exposed at this point.

For both the history server and the live UI, the goal is to show the peak value for each metric each executor. For the executors tab, this is the peak value of each metric over the lifetime of the executor. For the stages tab, this is the peak value for each metric for that executor while the stage is running. The executor could be processing tasks for other stages as well, if there are concurrent stages, or no tasks for this stage if it isn't assigned any tasks, but it is the peak values between the time the stage starts and ends.

Can you describe how the stage level metrics would work the last timestamp for any peak metric? Would there be a check to see if the event is being read from the history log?

Reply:
Just discussed with @squito -- thanks! Logging -1 for timestamp is confusing and hacky. 

Some items discussed:

For ExecutorMetrics, timestamp can be optional, or it can be removed completely and replaced by Array[Long], with comments explaining how the metrics work.

For logging, stage ID could be added as part of the executorMetrics in SparkListenerExecutorMetricsUpdate, but this is awkward, since this information isn't used as part of the Heartbeat, and only for logging. While the application is running, there could be multiple stages running when the metrics are gathered, so specifying 1 stage ID doesn't make sense. For logging, the metrics are the peak values for a particular stage, so are associated with a stage.

Another option is to add the information to SparkListenerStageCompleted, but this would bloat the event if there are many executors.

A third option is to create a new event, SparkListenerStageExecutorMetrics, which would have the executor ID, stage ID and attempt, and peak metrics. 

I'll give the 3rd option a try, and will add details to the design doc once this is more finalized.